tests/test_dag.py

```diff
@@ -418,3 +428,193 @@ def done_callback_test(dag):
+        self.assertEqual(
+            "some_string", internal._replace_nodes_with_results("some_string")
+        )
+        self.assertEqual(b"bytes", internal._replace_nodes_with_results(b"bytes"))
+
+    def test_self_recursion(self):
+        dct = {
+            "nod": _node(("a", "b", "c")),
+        }
+        lst = [dct, _node("nod")]
+        dct["lst"] = lst
+        tup = (_node("a lst:"), lst, _node("a dct:"), dct)
+        lst.append(tup)
+        got_lst = internal._replace_nodes_with_results(lst)
+
+        want_dct = {
+            "nod": ("a", "b", "c"),
+        }
+        want_lst = [want_dct, "nod"]
+        want_dct["lst"] = want_lst
+        want_tup = ("a lst:", want_lst, "a dct:", want_dct)
+        want_lst.append(want_tup)
+
+        # assertEqual will blow up if you try to compare a recursive structure.
+        # However, pickle handles recursion, so we just see if the results
+        # have the same pickling.
+        self.assertEqual(pickle.dumps(want_lst), pickle.dumps(got_lst))
+
+    def test_custom_types(self):
+        my_seq = CustomSeq("abcde")
+        my_seq.append(CustomSeq(("F", _node("G"))))
+        my_seq.append(_node("h"))
+
+        self.assertEqual(
+            CustomSeq(("a", "b", "c", "d", "e", CustomSeq("FG"), "h")),
+            internal._replace_nodes_with_results(my_seq),
+        )
+
+        self.assertEqual(
+            collections.OrderedDict(zero=0, one=1, two=2),
+            internal._replace_nodes_with_results(
+                collections.OrderedDict(
+                    zero=_node(0),
+                    one=_node(1),
+                    two=2,
+                )
+            ),
+        )
+
+    def test_dont_enter_special_types(self):
+        df = pd.DataFrame([{"a": 1, "b": 2}, {"a": 2, "b": 3}])
+        self.assertIs(df, internal._replace_nodes_with_results(df))
+        arr = np.array([["a", 1], ["b", 2]])
+        self.assertIs(arr, internal._replace_nodes_with_results(arr))
+
+    def test_ref_equality(self):
+        lst = ["a", _node("B"), "c"]
+        got_val = internal._replace_nodes_with_results([lst, lst])
+        self.assertIs(got_val[0], got_val[1])
+        self.assertIsNot(lst, got_val[0])
+
+    def test_replace_stored_params(self):
+        # Since the corner cases of tree traversal are tested in the other
+        # ReplaceNodesTests, this just tests the basics of stored params.
+        got = internal.replace_stored_params(
+            [
+                sp.StoredParam(_uid(0x1), decoder=results.Decoder("json")),
+                {
+                    "sub-dict": sp.StoredParam(
+                        _uid(0x2), decoder=results.Decoder("native")
+                    )
+                },
+                (sp.StoredParam(_uid(0x1), decoder=results.Decoder("json")),),
+            ],
+            sp.ParamLoader(
+                {
+                    "00000000-0000-0000-0000-000000000001": "W3RydWUsICJ0d28iLCAzXQ==",
```

Comment:
(also, should the data be a bytes object `b"..."`?)

Reply:
Since the data is loaded from JSON, it is going to be a string.

Also started base64ing locally to make it easier to understand what we're unserializing.